articles/build-a-restful-api-with-adonisjs/index.md

```diff
@@ -0,0 +1,419 @@
+                }
+                return; // 422
+            }
+            return; // 401
+        }
+
+        public async store({ auth request, response}: HttpContextContract)
+        {
+            const user = await auth.authenticate();
+            const post = new Post();
+            post.title = request.input('title');
+            post.content = request.input('content');
+            post.forumId = request.input('forum');
+            await user.related('posts').save(post)
+            return post
+        }
+        public async destroy({response, auth, request, params}: HttpContextContract)
+        {
+           const user = await auth.authenticate();
+           const post = await Post.query().where('user_id', user.id).where('id', params.id).delete();
+           return response.redirect('/dashboard');
+        }
+    }
+```
+
+You can [Clone the repository](https://github.com/Kaperskyguru/adonisjs-forum-api) to review the other controllers that I have created.
+
+### Creating Endpoint Routes
+The next step is to create the endpoints for our frontend or mobile App to consume.
+
+To to that, open the `route.ts` file inside `start` folder and add the following code.
+
+```ts
+//......
+Route.group(() => {
+  Route.post("register", "AuthController.register");
+  Route.post("login", "AuthController.login");
+  Route.group(() => {
+    Route.resource("posts", "PostsController").apiOnly();
+    Route.resource("forums", "ForumsController").apiOnly();
+    Route.get("users/forums", "UsersController.forumsByUser");
+    Route.get("users/posts", "UsersController.postsByUser");
+  }).middleware("auth:api");
+}).prefix("api");
+
+//......
+```
+
+So far, we have created different endpoints for our Forum API.
+
+Note that the `resource` method will create all the CRUD endpoints we need for our Forum API once. You can review more about it [here](https://preview.adonisjs.com/guides/http/routing#crud-actions).
+
+### Testing the Forum API
+When testing your endpoint using any HTTP clients e.g. [Postwoman](http://postwoman.io/), You might be faced with an error saying `Cannot find module 'phc-argon2'` when testing, please run this command to install the package.
+
+```bash
+    npm install phc-argon2
+```
+
+If we test the `posts` endpoint without authentication, we will be faced with the error below.
+
+![Test API Error](/engineering-education/build-a-restful-api-with-adonisjs/test-api-error.png)
+
+But if we login using the `/api/login` or register using the `/api/register` endpoint to retrieve our API Token:
+
+![Test API Auth](/engineering-education/build-a-restful-api-with-adonisjs/test-api-auth.png)
+
+We can access the protected endpoints after inserting the token as the Authorization header value.
+
+![Test API Header](/engineering-education/build-a-restful-api-with-adonisjs/test-api-header.png)
+
+Now, we can access our protected endpoints:
+
+![Test API](/engineering-education/build-a-restful-api-with-adonisjs/test-api.gif)
```

Comment:
The gif seems rather long - `about 25 seconds`. @hectorkambow advised it should be uploaded on YouTube and link embedded on the article. You can learn from the docmentation.

Reply:
done